Content.Server/Light/EntitySystems/HandheldLightSystem.cs

```diff
@@ -0,0 +1,284 @@
+        [Dependency] private readonly PopupSystem _popup = default!;
+
+        // TODO: Ideally you'd be able to subscribe to power stuff to get events at certain percentages.. or something?
+        // But for now this will be better anyway.
+        private HashSet<HandheldLightComponent> _activeLights = new();
+
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            SubscribeLocalEvent<HandheldLightComponent, ComponentInit>(OnInit);
+            SubscribeLocalEvent<HandheldLightComponent, ComponentRemove>(OnRemove);
+            SubscribeLocalEvent<HandheldLightComponent, ComponentGetState>(OnGetState);
+
+            SubscribeLocalEvent<HandheldLightComponent, ExaminedEvent>(OnExamine);
+            SubscribeLocalEvent<HandheldLightComponent, GetActivationVerbsEvent>(AddToggleLightVerb);
+            SubscribeLocalEvent<HandheldLightComponent, InteractUsingEvent>(OnInteractUsing);
+            SubscribeLocalEvent<HandheldLightComponent, UseInHandEvent>(OnUse);
+        }
+
+        private void OnGetState(EntityUid uid, HandheldLightComponent component, ref ComponentGetState args)
+        {
+            args.State = new SharedHandheldLightComponent.HandheldLightComponentState(GetLevel(component));
+        }
+
+        private byte? GetLevel(HandheldLightComponent component)
+        {
+            // Curently every single flashlight has the same number of levels for status and that's all it uses the charge for
+            // Thus we'll just check if the level changes.
+            if (component.Cell == null)
+                return null;
+
+            var currentCharge = component.Cell.CurrentCharge;
+
+            if (MathHelper.CloseToPercent(currentCharge, 0) || component.Wattage > currentCharge)
+                return 0;
+
+            return (byte?) ContentHelpers.RoundToNearestLevels(currentCharge / component.Cell.MaxCharge * 255, 255, SharedHandheldLightComponent.StatusLevels);
+        }
+
+        private void OnInit(EntityUid uid, HandheldLightComponent component, ComponentInit args)
+        {
+            EntityManager.EnsureComponent<PointLightComponent>(uid);
+            component.CellSlot = EntityManager.EnsureComponent<PowerCellSlotComponent>(uid);
+
+            // Want to make sure client has latest data on level so battery displays properly.
+            component.Dirty(EntityManager);
+        }
+
+        private void OnRemove(EntityUid uid, HandheldLightComponent component, ComponentRemove args)
+        {
+            HandleDeactivate(component);
+        }
+
+        private void OnInteractUsing(EntityUid uid, HandheldLightComponent component, InteractUsingEvent args)
+        {
+            if (args.Handled) return;
+
+            if (!_blocker.CanInteract(args.User)) return;
+            if (!component.CellSlot.InsertCell(args.Used)) return;
+            component.Dirty(EntityManager);
+            args.Handled = true;
+        }
```

Comment:
Eventually this interaction should be generalised to work with any `PowerCellSlotComponent` but it's fine for now I guess.

Reply:
I think there was a PR working on it but it ended up derelict.

Reply:
I've starting working on that, inc making the battery barrel component just use PowerCellSlots. So would be nice to get this merged to avoid conflicts.